NetKet/Graph/custom_graph.hpp

```diff
@@ -55,7 +81,41 @@ class CustomGraph : public AbstractGraph {
       : adjlist_(adjl), isbipartite_(isbipartite) {
     Init(size, edges, automorphisms, edgecolors);
   }
+#endif
 
+  CustomGraph(std::vector<Edge> edges, ColorMap colors = ColorMap(),
+              std::vector<std::vector<int>> automorphisms =
+                  std::vector<std::vector<int>>(),
+              bool isbipartite = false)
+      : edges_{std::move(edges)},
+        eclist_{std::move(colors)},
+        automorphisms_{std::move(automorphisms)},
+        isbipartite_{isbipartite} {
+    nsites_ = CheckEdges();
+    if (nsites_ == 0) {
+      throw std::invalid_argument{"Empty graphs are not supported."};
```

Comment:
If you throw `netket::InvalidInputError` it will be caught by the code in `netket.cc`.

Reply:
There's no `netket.cc` anymore -- we're moving to Python. That's why I use `std::invalid_argument` which is correctly converted to `ValueError` by pybind11.

Reply:
> True, but check if throwing in python has issues with Mpi etc.

@gcarleo could you elaborate? I'm afraid I don't get your point.

Reply:
@twesterhout We can [convert](https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#registering-custom-translators) `InvalidInputError` to a Python `ValueError`, which we probably should do.
I do not have particularly strong feelings about whether to use custom exceptions or the `std` ones, but it should be consistent throughout the project.

Reply:
@femtobit indeed we can.
Using `std` exceptions here is the first step in getting rid of NetKet's custom exceptions entirely since they add no additional info and are non-conformant (i.e. throwing them is most likely UB, but I'm not sure about the UB part).

Reply:
At least the C++ Core guidelines seem to disagree [(see E.14)](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#e14-use-purpose-designed-user-defined-types-as-exceptions-not-built-in-types). (Looking at this, I think it would be better for our custom type to inherit from `std::logic_error`, just as `std::invalid_argument` does, however.) Though I guess this is mostly a stylistic question for us, except of course for your point regarding potential UB.
I'd be interested to hear why you suspect throwing them to be UB and how this should be fixed.

Reply:
Guys, let's just use some ```netket::whatever_exception``` and define ```using netket::whatever_exception=whatever_we_will_decide```. The natural place do this is in the `Utils/exceptions.hpp`

Reply:
> At least the C++ Core guidelines seem to disagree (see E.14)

@femtobit touché! The reason it's not really applicable here is that all exceptions are converted to Python analogues anyway... So let's stick to @gcarleo's suggestion.

> I'd be interested to hear why you suspect throwing them to be UB and how this should be fixed.

As per [18.8.2] (taken from C++17 draft I have at hand, but I'm pretty sure this didn't change since C++11):
> Each standard library class T that derives from class exception shall have a publicly accessible copy constructor and a publicly accessible copy assignment operator that do not exit with an exception.

NetKet's exceptions store a `std::string`. `std::string`'s copy constructor and assignment operators allocate memory and can thus throw. That's why exceptions in `std` are usually implemented using COW (copy-on-write) strings.